<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="canonical" href="https://mrcieu.github.io/gwas2vcf/downstream/">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Downstream analyses - gwas2vcf</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Downstream analyses";
    var mkdocs_page_input_path = "downstream.md";
    var mkdocs_page_url = "/gwas2vcf/downstream/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> gwas2vcf</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../install/">Install</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../tutorial/">Tutorial</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">Downstream analyses</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#data">Data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#parsing-libraries">Parsing libraries</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#command-line-manipulation">Command-line manipulation</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#filter">Filter</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#extract-genome-wide-significant-variants">Extract genome-wide significant variants</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#extract-variants-by-gene">Extract variants by gene</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#extract-variants-by-pathway">Extract variants by pathway</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#select-genome-region-for-further-analysis">Select genome region for further analysis</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#annotate">Annotate</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#add-variant-frequency">Add variant frequency</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#add-genepathway-annotations">Add gene/pathway annotations</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#convert">Convert</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#export-to-nhgri-ebi-gwas-catalog-format">Export to NHGRI-EBI GWAS catalog format</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#liftover">Liftover</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#map-genomic-coordinates-to-another-genome-build-liftover">Map genomic coordinates to another genome build (liftover)</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#merge">Merge</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#combine-multiple-gwas-summary-stats-into-a-single-file">Combine multiple GWAS summary stats into a single file</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#validate">Validate</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#check-the-file-format-is-valid">Check the file format is valid</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">gwas2vcf</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Downstream analyses</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>

          <div role="main">
            <div class="section">
              
                <h1 id="working-with-gwas-vcf">Working with GWAS-VCF</h1>
<p>See below examples of working with GWAS-VCF. Let us know if you have other use cases through the issues page!</p>
<h2 id="data">Data</h2>
<p>Full GWAS summary statistics harmonised to GWAS-VCF are available on &gt;14,000 datasets available from <a href="https://gwas.mrcieu.ac.uk/" target="_blank">https://gwas.mrcieu.ac.uk/</a></p>
<h2 id="parsing-libraries">Parsing libraries</h2>
<p>See <a href="https://github.com/mrcieu/gwasvcf" target="_blank">R</a> and <a href="https://github.com/mrcieu/pygwasvcf" target="_blank">Python3</a> libraries for reading GWAS summary statistics in GWAS-VCF</p>
<h2 id="command-line-manipulation">Command-line manipulation</h2>
<p>The following examples require:</p>
<ul>
<li><a href="http://samtools.github.io/bcftools/">bcftools</a></li>
<li><a href="https://gatk.broadinstitute.org/">gatk</a></li>
<li><a href="http://www.htslib.org/download/">htslib</a></li>
</ul>
<p>Please cite the relevant tool(s) if you use these examples.</p>
<h3 id="filter">Filter</h3>
<h4 id="extract-genome-wide-significant-variants">Extract genome-wide significant variants</h4>
<p>The LP field is -log10(P), 7.3 is approx 5e-8</p>
<pre><code class="language-sh">bcftools filter \
-i 'FORMAT/LP &gt; 7.3' \
-o output.vcf \
file.vcf.gz
</code></pre>
<h4 id="extract-variants-by-gene">Extract variants by gene</h4>
<p>Requires annotation by Ensembl (see below)</p>
<pre><code class="language-sh">bcftools filter \
-i 'INFO/ENSG_ID == &quot;ENSG00000198670&quot;' \
file.vcf.gz
</code></pre>
<h4 id="extract-variants-by-pathway">Extract variants by pathway</h4>
<p>Requires annotation by Reactome (see below)</p>
<pre><code class="language-sh">bcftools filter \
-i 'INFO/Reactome_ID == &quot;R-HSA-3000171&quot;' \
file.vcf.gz
</code></pre>
<h4 id="select-genome-region-for-further-analysis">Select genome region for further analysis</h4>
<pre><code class="language-sh">bcftools filter \
-r 1:1000000-2000000 \
-o output.vcf.gz \
input.vcf.gz
</code></pre>
<h3 id="annotate">Annotate</h3>
<h4 id="add-variant-frequency">Add variant frequency</h4>
<pre><code class="language-sh"># download 1000 genomes phase 3 (hg19/GRCh37) allele frequencies and index
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz.tbi 

# annotate GWAS-VCF with 1kg allele frequencies
bcftools annotate \
-a ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz \
-c AF \
-O z \
-o output.vcf.gz \
input.vcf.gz
</code></pre>
<h4 id="add-genepathway-annotations">Add gene/pathway annotations</h4>
<p>Download and merge input files</p>
<pre><code class="language-sh"># download ensembl-to-position mapping and sort by ensgId
curl ftp://ftp.ensembl.org/pub/grch37/release-87/gff3/homo_sapiens/Homo_sapiens.GRCh37.87.gff3.gz | \
gzip -dc | \
awk -F&quot;\t|:|;|=&quot; '$3==&quot;gene&quot; &amp;&amp; $1 &gt;= 1 &amp;&amp; $1 &lt;= 22 {print $11&quot;\t&quot;$1&quot;\t&quot;$4&quot;\t&quot;$5}' | \
sort -k1,1 &gt; Ensembl2Position.sorted.txt

# download ensembl-to-pathway mapping and sort by ensgId
curl https://reactome.org/download/current/Ensembl2Reactome.txt | \
grep &quot;Homo sapiens&quot; | \
cut -s -f1,2 | \
sort -k1,1 &gt; Ensembl2Reactome.sorted.txt

# merge tables by ensgId, sort, compress and index
join \
-t $'\t' \
--check-order \
Ensembl2Position.sorted.txt \
Ensembl2Reactome.sorted.txt | \
awk -F&quot;\t&quot; '{print $2&quot;\t&quot;$3-1&quot;\t&quot;$4&quot;\t&quot;$5}' | \
sort -k1,1V -k2,2n -k3,3n | \
bgzip -c &gt; reactome.bed.gz
tabix -p bed reactome.bed.gz

# sort, compress and index ensembl-to-position mapping
awk -F&quot;\t&quot; '{print $2&quot;\t&quot;$3-1&quot;\t&quot;$4&quot;\t&quot;$1}' Ensembl2Position.sorted.txt | \
sort -k1,1V -k2,2n -k3,3n  | \
bgzip -c &gt; ensembl.bed.gz
tabix -p bed ensembl.bed.gz
</code></pre>
<p>Annotate GWAS-VCF with gene ID</p>
<pre><code class="language-sh"># annotate GWAS-VCF
bcftools annotate \
-a ensembl.bed.gz \
-c CHROM,FROM,TO,ENSG_ID \
-h &lt;(echo '##INFO=&lt;ID=ENSG_ID,Number=.,Type=String,Description=&quot;Ensembl gene ID&quot;&gt;') \
-o output.vcf.gz \
-O z \
-l ENSG_ID:unique \
input.vcf.gz
</code></pre>
<p>Annotate GWAS-VCF with Reactome ID</p>
<pre><code class="language-sh"># annotate GWAS-VCF
bcftools annotate \
-a reactome.bed.gz \
-c CHROM,FROM,TO,Reactome_ID \
-h &lt;(echo '##INFO=&lt;ID=Reactome_ID,Number=.,Type=String,Description=&quot;Reactome ID&quot;&gt;') \
-o output.vcf.gz \
-O z \
-l Reactome_ID:unique \
input.vcf.gz
</code></pre>
<h3 id="convert">Convert</h3>
<h4 id="export-to-nhgri-ebi-gwas-catalog-format">Export to NHGRI-EBI GWAS catalog format</h4>
<pre><code class="language-sh"># map to GWAS catalog format
bcftools query \
-e 'ID == &quot;.&quot;' \
-f '%ID\t[%LP]\t%CHROM\t%POS\t%ALT\t%REF\t%AF\t[%ES\t%SE]\n' \
gwas.vcf.gz | \
awk 'BEGIN {print &quot;variant_id\tp_value\tchromosome\tbase_pair_location\teffect_allele\tother_allele\teffect_allele_frequency\tbeta\tstandard_error&quot;}; {OFS=&quot;\t&quot;; if ($2==0) $2=1; else if ($2==999) $2=0; else $2=10^-$2; print}' &gt; gwas.tsv

# validate file using [ss-validate](https://pypi.org/project/ss-validate)
ss-validate -f gwas.tsv
</code></pre>
<h3 id="liftover">Liftover</h3>
<h4 id="map-genomic-coordinates-to-another-genome-build-liftover">Map genomic coordinates to another genome build (liftover)</h4>
<p>This procedure requires a chain file which contains the chromosome base-position mapping between two genome builds</p>
<pre><code class="language-sh"># download chain file
wget http://fileserve.mrcieu.ac.uk/ref/chains/b36tob37.chain
wget http://fileserve.mrcieu.ac.uk/ref/chains/b37tob36.chain
wget http://fileserve.mrcieu.ac.uk/ref/chains/b37tohg18.chain
wget http://fileserve.mrcieu.ac.uk/ref/chains/b37tohg19.chain
wget http://fileserve.mrcieu.ac.uk/ref/chains/hg18tob37.chain
wget http://fileserve.mrcieu.ac.uk/ref/chains/hg19toHg18.chain
curl https://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg38.over.chain.gz | zcat | sed 's/chr//' &gt; b37ToHg38.chain
</code></pre>
<pre><code class="language-sh"># perform liftover
 gatk LiftoverVcf \
--INPUT input.vcf.gz \
--OUTPUT output.vcf.gz \
--REJECT rejected.vcf.gz \
--CHAIN file.chain \
--REFERENCE_SEQUENCE target.fasta \
--RECOVER_SWAPPED_REF_ALT false 
</code></pre>
<h3 id="merge">Merge</h3>
<h4 id="combine-multiple-gwas-summary-stats-into-a-single-file">Combine multiple GWAS summary stats into a single file</h4>
<p>This is useful for distributing QTL/molecular phenotype GWAS</p>
<pre><code class="language-sh">bcftools merge \
-O z \
-o merged.vcf.gz \
*.vcf.gz
</code></pre>
<h3 id="validate">Validate</h3>
<h4 id="check-the-file-format-is-valid">Check the file format is valid</h4>
<pre><code class="language-sh">gatk ValidateVariants \
-V input.vcf.gz \
-R ref.fasta \
--validation-type-to-exclude ALLELES
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../tutorial/" class="btn btn-neutral" title="Tutorial"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../tutorial/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
